src/encoding.rs

```diff
@@ -690,75 +689,76 @@ macro_rules! length_delimited {
-                                                       encoded_len_repeated)
-                }
-            }
-        }
     };
 }
 
 pub mod string {
     use super::*;
 
-    pub fn encode<B>(tag: u32, value: &String, buf: &mut B)
+    pub fn encode<B>(tag: u32, value: &BytesString, buf: &mut B)
     where
         B: BufMut,
     {
         encode_key(tag, WireType::LengthDelimited, buf);
         encode_varint(value.len() as u64, buf);
-        buf.put_slice(value.as_bytes());
+        buf.put_slice(value.into());
     }
-    pub fn merge<B>(wire_type: WireType, value: &mut String, buf: &mut B) -> Result<(), DecodeError>
+    pub fn merge<B>(
+        wire_type: WireType,
+        value: &mut BytesString,
+        buf: &mut B,
+    ) -> Result<(), DecodeError>
     where
         B: Buf,
     {
         unsafe {
-            // String::as_mut_vec is unsafe because it doesn't check that the bytes
+            // BytesString::as_bytes_mut is unsafe because it doesn't check that the bytes
             // inserted into it the resulting vec are valid UTF-8. We check
             // explicitly in order to ensure this is safe.
-            super::bytes::merge(wire_type, value.as_mut_vec(), buf)?;
+
+            super::bytes::merge(wire_type, value.as_bytes_mut(), buf)?;
             str::from_utf8(value.as_bytes())
                 .map_err(|_| DecodeError::new("invalid string value: data is not UTF-8 encoded"))?;
         }
         Ok(())
     }
 
-    length_delimited!(String);
+    length_delimited!(BytesString);
+
+    #[cfg(test)]
+    mod test {
+        use quickcheck::{quickcheck, TestResult};
+
+        use super::super::test::{check_collection_type, check_type};
+        use super::*;
+
+        quickcheck! {
+            fn check(value: BytesString, tag: u32) -> TestResult {
+                super::test::check_type(value, tag, WireType::LengthDelimited,
+                                        encode, merge, encoded_len)
+            }
+            fn check_repeated(value: Vec<BytesString>, tag: u32) -> TestResult {
+                super::test::check_collection_type(value, tag, WireType::LengthDelimited,
+                                                   encode_repeated, merge_repeated,
+                                                   encoded_len_repeated)
+            }
+        }
+    }
 }
 
 pub mod bytes {
     use super::*;
 
-    pub fn encode<B>(tag: u32, value: &Vec<u8>, buf: &mut B)
+    pub fn encode<B>(tag: u32, value: &Bytes, buf: &mut B)
     where
         B: BufMut,
     {
         encode_key(tag, WireType::LengthDelimited, buf);
         encode_varint(value.len() as u64, buf);
-        buf.put_slice(value);
+        buf.put(value);
     }
 
-    pub fn merge<B>(
-        wire_type: WireType,
-        value: &mut Vec<u8>,
-        buf: &mut B,
-    ) -> Result<(), DecodeError>
+    pub fn merge<B>(wire_type: WireType, value: &mut Bytes, buf: &mut B) -> Result<(), DecodeError>
```

Comment:
In the current shape, this just replaces `Vec::extend_from_slice` with `Bytes::extend_from_slice`, so it does not really achieve the zero-copy optimization requested for by #31 on the decoding side. It's hard to see how this code alone could, though, since the source buffer is behind the `Buf` abstraction. How about changing the `merge` method to extract data from `Bytes`, so that the value representations backed by `Bytes` could split from the buffer?

Reply:
Right, this is why I've been holding off on adding support for `Bytes` fields; as I mentioned in #31 it's not possible to get the ref counting zero copy support without specialization.  Switching `Bytes` for `Buf` in the `Message` API is not something I want to do.

Reply:
To be clear, I don't have any issue landing a PR to add support for `Bytes` fields, and if it's technically possible I'm also in favor of supporting zero-copy decoding through a nightly feature flag + specialization.  The only thing I'm not willing to do at this point is change the `Message` API to enable zero copy.